L1Trigger/TrackFindingTracklet/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,261 @@
+          // cut on outer eta sector boundaries
+          if (binEta == -1 || binEta == setup_->numSectorsEta())
+            continue;
+          cot -= setup_->sectorCot(binEta);
+          zT -= setup_->sectorCot(binEta) * setup_->chosenRofZ();
+          // cut on eta and |z0| < 15 cm
+          if (!dfZT.inRange(zT) || !dfCot.inRange(cot))
+            continue;
+          const int binZT = dfZT.toUnsigned(dfZT.integer(zT));
+          const int binCot = dfCot.toUnsigned(dfCot.integer(cot));
+          const vector<int>& layerEncoding = layerEncoding_->layerEncoding(binEta, binZT, binCot);
+          // get rphi parameter
+          double qOverPt = ttTrackRef->rInv() / 2.;
+          // calculcate track phi at radius hybridChosenRofPhi with respect to phi sector centre
+          double phiT = deltaPhi(ttTrackRef->phi() - setup_->hybridChosenRofPhi() * qOverPt - ttTrackRef->phiSector() * setup_->baseRegion());
+          const int sectorPhi = phiT < 0. ? 0 : 1; // dirty hack
+          phiT -= (sectorPhi - .5) * setup_->baseSector();
+          // cut on nonant size and pt
+          if (!dfPhiT.inRange(phiT) || !dfQoverPt.inRange(qOverPt))
+            continue;
+          // loop over stubs
+          TTBV hitPattern(0, setup_->numLayers());
+          vector<int> layerMap(setup_->numLayers(), 0);
+          for (const TTStubRef& ttStubRef : ttTrackRef->getStubRefs()) {
+            const GlobalPoint& gp = setup_->stubPos(ttStubRef);
+            const double r = gp.perp() - setup_->hybridChosenRofPhi();
+            const double phi = deltaPhi(gp.phi() - (ttTrackRef->phi() - qOverPt * gp.perp()));
+            const double z = gp.z() - (ttTrackRef->z0() + ttTrackRef->tanL() * gp.perp());
+            double layer = distance(layerEncoding.begin(), find(layerEncoding.begin(), layerEncoding.end(), setup_->layerId(ttStubRef)));
+            if (layer >= setup_->numLayers())
+              layer = setup_->numLayers() - 1;
+            // cut on phi and z residuals
+            if (!dfPhi.inRange(phi) || !dfZ.inRange(z))
+              continue;
+            hitPattern.set(layer);
+            int& nLayerStubs = layerMap[layer];
+            // cut on max 4 stubs per layer
+            if (nLayerStubs == setup_->kfMaxStubsPerLayer())
+              continue;
+            nLayerStubs++;
+            numLayerStubs[layer]++;
+            stubs.emplace_back(ttStubRef, dataFormats_, r, phi, z, trackId, layer);
+          }
+          const TTBV& maybePattern = layerEncoding_->maybePattern(binEta, binZT, binCot);
+          tracks.emplace_back(ttTrackRef, dataFormats_, hitPattern, setup_->layerMap(hitPattern, layerMap), maybePattern, phiT, qOverPt, zT, cot, sectorPhi, binEta, trackId++);
+        }
+        // truncate and transform stubs
+        vector<vector<StubKFin*>> layerStubs(setup_->numLayers());
+        for (int layer = 0; layer < setup_->numLayers(); layer++)
+          layerStubs[layer].reserve(numLayerStubs[layer]);
```

Comment:
At several points in this class, lines of code are used to reserve space in vectors. Consider replacing the vectors by std::list, so reservation is no longer necessary.

Reply:
see my comment on lists above.